CTP-4656: using switch to only return missing marks for given marker for coursework assessments

opitz opened the pull request:
Making use of the newly introduced boolean switch in feedback_tracker::count_missing_grades() which allows to return only coursework missing grades for the current user as assessor.

opitz committed:
CTP-4656: using switch to only return missing marks for given marker for coursework assessments

leonstr commented on block_my_feedback.php:
```diff
@@ -325,7 +325,7 @@ public static function add_mod_data(cm_info $mod, stdClass $assess): bool {
         // Check that mod has missing markings.
         $submitterids = array_column(feedback_tracker::get_module_submissions($mod), 'userid');
-        if (!$assess->requiremarking = feedback_tracker::count_missing_grades($mod, $submitterids, $gradeitemid)) {
+        if (!$assess->requiremarking = feedback_tracker::count_missing_grades($mod, $submitterids, $gradeitemid, true)) {
```
I think this could be clearer. `count_missing_grades()` returns an int so we're checking if it's 0. And I don't like assignments in conditions when the variable is used outside the subsequent block (personal taste). So I would prefer:

```
        $assess->requiremarking = feedback_tracker::count_missing_grades($mod, $submitterids, $gradeitemid, true);

        if ($assess->requiremarking === 0) {
            return false;
        }
```

opitz replied:
refactored

opitz committed:
CTP-4656: show coursework feedback to students
- getting supported course module types from report_feedback_tracker
- addressed code review feedback

opitz committed:
installing report_feedback_tracker for tests

opitz committed:
reversed

leonstr commented on block_my_feedback.php:
```diff
@@ -456,7 +457,9 @@ public function get_submissions($user) {
         // Construct the IN clause.
-        list($insql, $params) = $DB->get_in_or_equal(['assign', 'quiz', 'turnitintooltwo'], SQL_PARAMS_NAMED);
+        // Get supported module types from feedback_tracker.
+        $supported = \report_feedback_tracker\local\helper::get_supported_types();
```
(In code review for Feedback Tracker I've suggested `get_supported_types()` be moved here since its only utility is here. Ignore this suggestion if you don't agree).

opitz replied:
I do agree. done now.

opitz committed:
transferred get_supported_types() from report_feedback_tracker

opitz committed:
improved get_supported_types()

leonstr reviewed (commented):
Between 6f02096 and db517ef we added report_feedback_tracker to `moodle-ci.yml`. Possibly this didn't work since report_feedback_tracker branch `main` did not have `get_supported_types()`? But since we've now moved this function to block_my_feedback and report_feedback_tracker `main` **does** have `is_supported_module()` possibly we should try re-instating this dependency in `moodle-ci.yml` since the tests are currently failing.

opitz commented:
@leonstr The issue is a bit more complicated: report_feedback_tracker is a internal - so for GitHub to be able to access that repo I will have to deposit a public key there - which I cannot due to insufficient access rights... I will have to ask @aspark21  to either give me the required rights or place the key for me... for now there is nothing I can do methinks...

leonstr commented:
In that case I think we should skip that call during testing, something like:

```
 21 use report_feedback_tracker\local\helper as feedback_tracker_helper; // UCL plugin.
    ⋮
277                     if (feedback_tracker_helper::is_supported_module($mod->modname) &&
    ⋮
533             if (array_key_exists($modname, $installed) && (PHPUNIT_TEST || feedback_tracker_helper::is_supported_module($modname))) {
```

It seems bad to have automated testing and break this. The red ×s everywhere are getting to me.

(Maybe any functions in `report_feedback_tracker\local\admin` that get called externally, e.g. from this plugin, should all be moved into `report_feedback_tracker\local\helper` – but let's ignore that for now!).

opitz committed:
support all module types for PHPUNIT_TEST

leonstr reviewed (approved):
Green ticks all round.
